### DISCUSSION TITLE
Error when installing virtual machine

### DISCUSSION THREAD

**Discussion by Verhoeckx** [cockpit-project/cockpit-machines#278] [General]
Hello Cockpit developers,

I successfully created a new virtual machine (with the name Nextcloud) but when I click on install, I get the following error message:
```
VM Nextcloud failed to get installed
ERROR Network not found: no network with matching name 'default' 
Domain installation does not appear to have been successful. If it was, you can restart your domain by running: 
virsh --connect qemu:///system start Nextcloud 
otherwise, please restart your installation.
```

I tried the command and got the following error:
```
error: Failed to start domain Nextcloud
error: Network not found: no network with matching name 'default'
```

Do I need to create some virtual network first?

These are the settings that I use:

![Screenshot 2021-07-27 at 15-36-23 Virtual Machines - verhoeckx IntelNUC](https://user-images.githubusercontent.com/64807887/127163628-dd5216b2-94bc-45e4-b7ec-315641f255a5.png)

![Screenshot 2021-07-27 at 15-37-10 Virtual Machines - verhoeckx IntelNUC](https://user-images.githubusercontent.com/64807887/127163632-eb6d8822-1ea5-454e-9274-73c6a48a3cfb.png)

In the first screenshot you can see that I want to install RHEL 8.4 but in the second screenshot it says RHEL 8.2?

Let me know what I can do.

**Comment by martinpitt** [comment]
The "default" system network should be provided by the `libvirt-daemon-config-network` package. Do you have this installed? What does `sudo virsh net-list --all` show?

**Reply by KKoukiou** [reply]
Thanks @martinpitt. In addition @Verhoeckx, in order that you define the default network you can just run:
`# virsh net-define /usr/share/libvirt/networks/default.xml`

If you really don't want the VM to use the default network, you can remove the NIC, before clicking install from clicking 'Remove' from the Networks card in the VM details page.

**Comment by Verhoeckx** [comment]
Hi @martinpitt and @KKoukiou!

This is what I did:

1. I installed the package `libvirt-daemon-config-network`.
2. I run the command `virsh net-define /usr/share/libvirt/networks/default.xml`.
3. I activated the default network in Cockpit.

When I now run `sudo virsh net-list --all` I get:
```
 Name      State    Autostart   Persistent
--------------------------------------------
 default   active   yes         yes
```

I'm documenting all the steps I make and will create I blog post about. Which of the above command were absolutely necessary and which don't? Should the package `libvirt-daemon-config-network` not have been installed when I installed Cockpit?

---

When I now try to install the virtual machine I get the following error:
`ERROR internal error: process exited while connecting to monitor: 2021-07-28T08:37:47.799278Z qemu-kvm: -blockdev {"driver":"file","filename":"/home/verhoeckx/rhel-8.4-x86_64-dvd.iso","node-name":"libvirt-1-storage","auto-read-only":true,"discard":"unmap"}: Could not open '/home/verhoeckx/rhel-8.4-x86_64-dvd.iso': Permission denied Domain installation does not appear to have been successful. If it was, you can restart your domain by running: virsh --connect qemu:///system start Nextcloud otherwise, please restart your installation.`

I guess Cockpit doesn't have permission to open the file `rhel-8.4-x86_64-dvd.iso`? Or is the domain installation still not successful?

**Reply by martinpitt** [reply]
I'm not sure about the network setup steps -- maybe that's also a missing dependency somewhere.

As for the permission error, this is usually an SELinux thing, check the journal for policy violations (grep for `avc.*denied`). I bet it works if you run `sudo setenforce 0`. Normally system libvirt only has permissions to access /var/lib/libvirt/images/, if you want it to access others, you need to `chcon` these. But for `/home` it's usually more appropriate to take the "session" libvirt connection instead of the "system" one? That also avoids problems with the system `default` network.

**Reply by KKoukiou** [reply]
Good point about the libvirt-daemon-config-network package possibly needed to be set as our dependency. 
In cockpit-machines we do depend on libvirt-daemon and libvirt-client, but none of these bring that package. @martinpitt does it make sense for you that we also specify it as a dependency?
Creating VMs on system connection are by default configured with the 'default' libvirt network, which indeed on a fresh install will be probably missing/

**Reply by martinpitt** [reply]
@KKoukiou : I'm not entirely sure. If it's cockpit-machines itself which makes the assumption about the default network, then it should depend on the package. If it's the generated XML from `virt-install`, then virt-manager should depend on it instead.

**Reply by Verhoeckx** [reply]
Quick questions:

- What is the difference between a system and a session connection? When should I choose what? The word session sounds like it's impermanent.
- Is moving the ISO image to the directory `/var/lib/libvirt/images/` also a good solution?

Thanks for helping me.

**Reply by Verhoeckx** [reply]
Strange, I suddenly also have a `default` storage pool with the exact same size as the `root` storage pool. The total amount of storage now exceeds the total size of the SSD.

**Reply by KKoukiou** [reply]
Session and system connection VMs have just different use cases. You need to take a look at the docs https://wiki.libvirt.org/page/FAQ#What_is_the_difference_between_qemu:.2F.2F.2Fsystem_and_qemu:.2F.2F.2Fsession.3F_Which_one_should_I_use.3F and decide what you need.

Regarding your ISO file location issue, you can indeed either move it to that directory that qemu can read, or you can create a storage pool pointing to your /home/yournamehere directory with correct permissions and selinux labels as seen here:

I am using the storage pool approach, but I really can't say what is the recommended way.

```
$ sudo virsh pool-dumpxml home 
<pool type='dir'>
  <name>home</name>
  <target>
    <path>/home/kkoukiou/</path>
    <permissions>
      <mode>0755</mode>
      <owner>1000</owner>
      <group>1000</group>
      <label>unconfined_u:object_r:user_home_t:s0</label>
    </permissions>
  </target>
</pool>

```

**Reply by Verhoeckx** [reply]
I moved the ISO image to the directory `/var/lib/libvirt/images/` and now the virtual machine starts fine.

According to the documentation it's best to use a `system` connection:
`You will definitely want to use qemu:///system if your VMs are acting as servers. VM autostart on host boot only works for 'system', and the root libvirtd instance has necessary permissions to use proper networkings via bridges or virtual networks.`

I added the paragraph [Install a default virtual network](https://jeroenverhoeckx.com/build-your-own-personal-cloud.html#install-a-default-virtual-network.html) to my blog post (this blog post is work in progress).

**Reply by Logikoz** [reply]
i solve this problem moving the .iso to /media folder.

**Comment by Verhoeckx** [comment]
I want to continue with the installation but I have the following problem.

I want to increase the size of the root volume because I want to create a virtual machine of 100 GB for the Nextcloud installation.
To be able to increase the size of the root volume I need to decrease the size of the home volume. However when I try to do that I get the message that XFS volumes can't be shrunk:

![Screenshot from 2021-07-28 13-52-15](https://user-images.githubusercontent.com/64807887/127317902-580782a0-e694-4bfc-bb2b-4a71f5deff98.png)

Now how can I change the size of the volumes?

---

And why does the virtual storage pools show three pools?
![Screenshot from 2021-07-28 14-04-30](https://user-images.githubusercontent.com/64807887/127319244-0fb783d2-edb9-46b0-adcd-1b942516ce8f.png)

**Comment by Verhoeckx** [comment]
According to [Red Hat official documentation](https://access.redhat.com/solutions/540013) XFS filesystems can't be shrunk :scream: !
I think the only solution is to do a complete new installation of RHEL? 

If XFS can't be shrunk then why does the installer (Anaconda) creates maximum volumes? This way they make it impossible to change the filesystem layout! This is really unbelievable. I can't call this professional software.

Shouldn't people be warned about this during the installation?

**Reply by garrett** [reply]
It's true; XFS filesystems cannot shrink, but can grow. (A lot of other filesystems can be shrunk, such as ext2/3/4, btrfs, ntfs, fat, etc... but XFS can only grow.)

Just some ideas:
- have a dedicated disk or partition elsewhere for your VMs?
- use the home partition (like `/home/virtualmachines/` or something similar) for storage, if there's enough space?
- back up the data on the home partition, remove the partition, and restore the data? (this could cause data loss if the data is not properly backed up or restored) — if you have enough space in the root filesystem, you could even temporarily copy the data (with permissions and extended attributes) into the root partition.

To be clear, I'm not suggesting you do any of the above; but they're possibly options you could look into for making sure you have enough space for your VM.

**Reply by Verhoeckx** [reply]
Hi @garrett,

Thanks for your ideas and feedback!


I played a little with different partition tables and ended up with the most simple one: one big `/` partition!
```
/dev/mapper/rhel-root   /                       xfs     defaults        0 0
UUID=ffda62b0-2abd-4edf-9352-7d7d8891419f /boot                   xfs     defaults        0 0
UUID=B968-8E11          /boot/efi               vfat    umask=0077,shortname=winnt 0 2
/dev/mapper/rhel-swap   none                    swap    defaults        0 0
```

I now use this partition table for the host and the virtual machine. To accomplish this I had to do a complete reinstall of the host and the virtual machine :scream: :laughing: !

I updated my documentation (Still work in progress):
[https://jeroenverhoeckx.com/build-your-own-personal-cloud.html](https://jeroenverhoeckx.com/build-your-own-personal-cloud.html)
